tokio/src/signal/unix.rs

```diff
@@ -432,28 +444,31 @@ impl Signal {
     /// }
     /// ```
     pub fn poll_recv(&mut self, cx: &mut Context<'_>) -> Poll<Option<()>> {
-        self.rx.poll_recv(cx)
-    }
-
-    /// Try to receive a signal notification without blocking or registering a waker.
-    pub(crate) fn try_recv(&mut self) -> Result<(), TryRecvError> {
-        self.rx.try_recv()
+        self.inner.poll_recv(cx)
     }
 }
 
 // Work around for abstracting streams internally
-pub(crate) trait InternalStream: Unpin {
+pub(crate) trait InternalStream {
     fn poll_recv(&mut self, cx: &mut Context<'_>) -> Poll<Option<()>>;
-    fn try_recv(&mut self) -> Result<(), TryRecvError>;
+    fn try_recv(&mut self) -> Option<()>;
 }
 
 impl InternalStream for Signal {
     fn poll_recv(&mut self, cx: &mut Context<'_>) -> Poll<Option<()>> {
         self.poll_recv(cx)
     }
 
-    fn try_recv(&mut self) -> Result<(), TryRecvError> {
-        self.try_recv()
+    fn try_recv(&mut self) -> Option<()> {
```

Comment:
Maybe the watch channel should be over a `u64` representing the number of sent signals. Then `try_recv` becomes "borrowing" the watch value and see if the value != the last seen value.

Reply:
I ended up saving a `usize` by reusing the version tracked within a `Receiver` itself